framework/src/transfers/MultiAppNearestNodeTransfer.C

```diff
@@ -111,10 +112,32 @@ MultiAppNearestNodeTransfer::execute()
 
       if (is_nodal)
       {
-        MeshBase::const_node_iterator node_it = to_mesh->local_nodes_begin();
-        MeshBase::const_node_iterator node_end = to_mesh->local_nodes_end();
+        std::vector<Node *> target_local_nodes;
 
-        for (; node_it != node_end; ++node_it)
+        if (isParamValid("target_boundary"))
+        {
+          BoundaryID target_bnd_id = _to_meshes[i_to]->getBoundaryID(getParam<BoundaryName>("target_boundary"));
+
+          ConstBndNodeRange & bnd_nodes = *(_to_meshes[i_to])->getBoundaryNodeRange();
+          for (ConstBndNodeRange::const_iterator nd = bnd_nodes.begin(); nd != bnd_nodes.end(); ++nd)
+          {
+            const BndNode * bnode = *nd;
+            if (bnode->_bnd_id == target_bnd_id && bnode->_node->processor_id() == processor_id())
+              target_local_nodes.push_back(bnode->_node);
+          }
+        }
+        else
+        {
+          target_local_nodes.resize(to_mesh->n_local_nodes());
```

Comment:
Hmm... store a list of node IDs instead?  I don't know all the details, but copying some pointers can't be that expensive, can it?  I'd worry about it if it showed up in profiling...